architecture.md

```diff
@@ -0,0 +1,122 @@
+#  Verifiable Credential Issuer HTTP API Architecture Model
+This document describes the architecture model and context for the Verifiable
+Credential Issuer HTTP API.
+
+## Overview
+The Verifiable Credential Issuer HTTP API is designed to be a consistent API for
+Issuer Services to call in order to generate Verifiable Credentials. This API is
+intended for Issuer Services who are able to manage and secure external
+connections to Holders. This API encapsulates and abstracts the Verifiable
+Credential data model, format, and proof methods from Issuer Services. This API
+also abstracts any underlying DID methods, resolver capabilities, and
+cryptographic algorithms from the Issuer Services.
+
+## Objectives
+The following capture the objectives of this API.
+1. Provide an open and consistent HTTP based API for Issuer Services to leverage
+when generating Verifiable Credentials.
+1. Abstract the underlying data model, formats and proof mechanisms from the
+Issuer Services.
+1. Abstract the underlying DID methods and cryptographic algorithms from the
+Issuer Services as required.
+1. Rely on 'upstream' Issuer Services to manage connections with external
+entities (like Holder software).
+
+These objectives allow for Issuer Services to modify their underlying Verifiable
+Credential, supported DID methods, and cryptographic suites software with
+minimal impact to their application services, business logic and user experience.  
+
+## Component Overview
+The following diagram is a high level, generic overview of the intended
+architecture context for this API.
+
+![Architecture Diagram](images/vc-issuer-http-api.png)
+
+### Components
+#### 1. User Agent
+This is the software the user (or individual) is using to interact with the
```

Comment:
Holders aren't always individuals. They could be IoT devices or institutions. For this reason, I don't believe the term "user agent" is a good one. The Hyperledger Indy and Aries communities avoided the phrase "user agent" for another reason, which is because it has a fairly well understood meaning in web browser land that is at odds with what we want it to mean (browsers are hardly agents of end users anymore; mostly they are agents of browser vendors). But set that all aside if you like; the main thing I wanted to say here is that I think "(or individual)" ought to be removed.

Reply:
But in this case we do mean user agent - something that interacts with a user (which then kicks over to an Agent/Holder to complete the flow). I can remove the "individual", that's fine.
Happy to accommodate other Things - like is it relevant to include the interfaces Things may use to bootstrap their Agents?

Reply:
My point is that the thing this agenty entity interacts with is often not a "user" -- but rather an IoT device, or some subset of an institution. I don't think either an IoT device or a subset of an institution qualifies as a "user."

Anyway, I guess it doesn't matter that much. You can resolve this concern.

Reply:
@dhh1128 I think @mavarley is using this definition: https://en.wikipedia.org/wiki/User_agent ... which is aligned with the way that W3C and IETF use the term user agent... from the article:

> In many cases, a user agent acts as a client in a network protocol used in communications within a client–server distributed computing system. In particular, the Hypertext Transfer Protocol (HTTP) identifies the client software originating the request, using a user-agent header, even when the client is not operated by a user. The Session Initiation Protocol (SIP) protocol (based on HTTP) followed this usage. In the SIP, the term user agent refers to both end points of a communications session.

Usually, the argument goes like this: at the very, very end of the interaction is a user of some kind, and that's why it's called a "user agent"... because, in the end, it's typically acting on behalf of a user (where the user could be person, AI, or corporation). I don't know if that helps or hurts things. :)

Reply:
If you follow the link on the word "user" in that wikipedia article, you get this as the first sentence on "user":

>A user is a person who utilizes a computer or network service

That's what's making me uncomfortable. Every holder is not a person. ("at the very, very end of the interaction is a user of some kind" is kind of like saying, "In the very long run, we're all dead"; it's true, but not particularly insightful or helpful.) The architectural component that this "user agent" is supposed to call should not be imagined as a purely UI-mediated construct built for people. I also don't like it because "user agent" is a very webby term. VCs are not an inherently webby thing.

I don't care about this issue so much that I'll resist any further. I just think this it is a term that perpetuates a bad mental model.

Reply:
After I posted my previous comment, I realized that it might have come across as being Brusque. When I said that the label user agent wasn't insightful, what I meant is that every piece of software on our diagram eventually has a user behind it at the very very end. So the user agent label doesn't distinguish one component from another in a helpful way I think the best label for this component would simply be client.